apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabUpgradeOpt/upOpt.ts

```diff
@@ -0,0 +1,641 @@
+      slotKey: art.slotKey,
+      mainStat: art.mainStatKey,
+      subs: art.substats
+        .map(({ key }) => key)
+        .filter((v) => v !== '') as SubstatKey[],
+      values: {
+        [art.setKey]: 1,
+        [art.mainStatKey]: toDecimal(art.mainStatKey, mainStatVal),
+        ...Object.fromEntries(
+          art.substats
+            .filter(({ key }) => key !== '')
+            .map((substat) => [
+              substat.key,
+              toDecimal(substat.key, substat.accurateValue),
+            ])
+        ), // Assumes substats cannot match main stat key
+      },
+    })
+  }
+
+  /** Calcs all artifacts using Fast method */
+  calcFastAll() {
+    function score(a: UpOptArtifact) {
+      return a.result!.p * a.result!.upAvg
+    }
+    this.artifacts.forEach((_, i) => this.calcFast(i, this.calc4th))
+    this.artifacts.sort((a, b) => score(b) - score(a))
+    this.fixedIx = 0
+  }
+
+  calcSlowToIndex(ix: number, lookahead = 5) {
+    const fixedList = this.artifacts.slice(0, this.fixedIx)
+    const arr = this.artifacts.slice(this.fixedIx)
+
+    function score(a: UpOptArtifact) {
+      return a.result!.p * a.result!.upAvg
+    }
+    function compare(a: UpOptArtifact, b: UpOptArtifact) {
+      if (score(a) > 1e-5 || score(b) > 1e-5) return score(b) - score(a)
+
+      const meanA = a.result!.distr.gmm.reduce(
+        (pv, { phi, mu }) => pv + phi * mu,
+        0
+      )
+      const meanB = b.result!.distr.gmm.reduce(
+        (pv, { phi, mu }) => pv + phi * mu,
+        0
+      )
+      return meanB - meanA
+    }
+
+    // Assume `fixedList` is all slowEval'd.
+    let i = 0
+    const end = Math.min(ix - this.fixedIx + lookahead, arr.length)
+    do {
+      for (; i < end; i++) this.calcSlow(this.fixedIx + i, this.calc4th)
+
+      arr.sort(compare)
+      this.artifacts = [...fixedList, ...arr]
+      for (i = 0; i < end; i++) {
+        if (arr[i].result!.evalMode === ResultType.Fast) break
+      }
+    } while (i < end)
+  }
+
+  /* Fast distribution to result. */
+  toResult1(
+    distr: { prob: number; mu: number[]; cov: number[] }[]
+  ): UpOptResult {
+    let ptot = 0
+    let upAvgtot = 0
+    const gmm = distr.map(({ prob, mu, cov }) => {
+      const z = mu.map((mui, i) => {
+        return gaussianPE(mui, cov[i], this.thresholds[i])
+      })
+      const p = Math.min(...z.map(({ p }) => p))
+      const cp = Math.min(...z.slice(1).map(({ p }) => p))
+      ptot += p * prob
+      upAvgtot += p * prob * z[0].upAvg
+      return { phi: prob, cp, mu: mu[0], sig2: cov[0] }
+    })
+    const lowers = gmm.map(({ mu, sig2 }) => mu - 4 * Math.sqrt(sig2))
+    const uppers = gmm.map(({ mu, sig2 }) => mu + 4 * Math.sqrt(sig2))
+    return {
+      p: ptot,
+      upAvg: ptot < 1e-6 ? 0 : upAvgtot / ptot,
+      distr: { gmm, lower: Math.min(...lowers), upper: Math.max(...uppers) },
+      evalMode: ResultType.Fast,
+    }
+  }
+
+  calcFast(ix: number, calc4th = true) {
```

Comment:
Maybe briefly doc in their respective functions, the difference between`_calcFast*`, `_calcSlow*`, `_calcExaact*`, and `_toResult*` variants.